### PR TITLE
Overhaul

### PR DESCRIPTION
**Pull request by XedinUnknown**
Contains many BC-breaking changes.

### CODE REVIEW

File: src/EventInterface.php [hunk 1]
```diff
@@ -1,14 +1,54 @@
...
+    public function getParams(): array;
+
+    /**
+     * Get a single parameter by name.
+     *
+     * @param string  $name    The name of the parameter to get.
+     * @param mixed   $default The value to return if the parameter does not exist.
+     *
+     * @return mixed The value of the parameter.
+     * @since [*next-version*]
```
Comment (mecha) [inline src/EventInterface.php]: The order of the tags here is inconsistent.

File: src/EventFactoryInterface.php [hunk 2]
```diff
@@ -1,23 +1,23 @@
...
  * A factory of events.
  *
  * @since [*next-version*]
  */
-interface EventFactoryInterface extends FactoryInterface
+interface EventFactoryInterface
 {
     /**
-     * {@inheritdoc}
+     * Creates a new event
```
Comment (mecha) [inline src/EventFactoryInterface.php]: Missing period.

File: src/EventInterface.php [hunk 3]
```diff
@@ -1,14 +1,54 @@
...
-interface EventInterface extends BaseEventInterface
+interface EventInterface
 {
+    /**
+     * Get event name.
+     *
+     * @since [*next-version*]
+     *
+     * @return string The name of the event.
+     *                The event name contains mixed-case alphanumeric characters, as well as '_' (underscore) and '.' (period).
```
Comment (mecha) [inline src/EventInterface.php:18]: What about `'/'`?

**Review by mecha — COMMENTED** [review]
Looks good. Just a few nitpicks related to PHP docs, and 1 question.

File: src/EventInterface.php (hunk 3 above)
Comment (XedinUnknown) [reply src/EventInterface.php:18]: This spec is taken directly from the previous draft PSR-14 version that we had mirrored and you had implemented. Do you think we should include the `'/'` there? I'd say, let's leave it for now as it was, and add if needed. I would like to think that FIG excluded this character for a reason.

Comment (mecha) [reply src/EventInterface.php:18]: Agreed. I only asked in case this _wasn't_ from PSR-14. Let's leave it for now.

**Comment by XedinUnknown** [comment]
So, uh, apparently I cannot have events be immutable, as pointed out by @widoz. The [PSR-14][] spec explicitly makes this impossible. The only way to pass data from one listener to another with immutable events would be to return the new events with changed data. But the spec explicitly says:

> A Dispatcher MUST ignore return values from Listeners.

I will therefore now fix the problem by replacing `withParams()` with `setParams()`.

I have to say, I'm pretty upset about having to make this change. In my humble opinion, events should be immutable always. Because if an event object has been emitted, it should remain that way forever. Let's imagine you keep track of all events that happened, for example because you want to use Event Sourcing in your application. So, once an event has been emitted, you push in onto a stack. If the event is mutable, then anything with a pointer to it can change it. But events cannot be changed once they have happened. I woke up this morning at 9:24, and nothing can undo that, nothing can change the circumstances of what has already happened in the past.

So, I have to say, FIG doin me a such big disappoint here. And this is coming from me, le PSR fanboye.

The only thing that I can do in order to help prevent mutation of emitted events is to declare an `EventInterface` and a `MutableEventInterface`. This way, although consumers may rely on a concrete implementation, they will never be type-safe that way, and this will discourage bad behaviour.

For FIG's reasoning, please see the [meta document][immutable-events].

[PSR-14]: https://www.php-fig.org/psr/psr-14/
[immutable-events]: https://www.php-fig.org/psr/psr-14/meta/#43-immutable-events

**Review by mecha — APPROVED** [review]
Honestly, I don't like the direction this has gone in, but at the same time I think this is as good as it can get. FIG are concerned with interoperability between frameworks, while here at Dhii we try to forge good SOLID standard packages.

Ultimately, I can say that this LGTM. But please note that I say that with a pinch of indifference. I can't see myself ever utilizing any such event/dispatch standard, just like how I can't see myself using any factory standard.

**Comment by XedinUnknown** [comment]
> I can't see myself ever utilizing any such event/dispatch standard

Is this because of the mutability of events?

**Comment by mecha** [comment]
No. It bothers me, but I can live with it. It's actually a couple of reasons:

1. The event interface is way too vague and loose, just like how our old `FactoryInterface` was. It promises nothing, provides nothing and puts burden on the consumer to do a bunch of sanity checks before working with it. I've come to see generic events in the same light. The only thing it promises is that it has data, and that data can be changed (or used to spawn a new instance). They're basically just extended containers. And just like containers, there's no guarantee that it has the data you need, putting additional burden on the listener.

2. I don't like the fact that to register events, I need to `addListener()`. When I see APIs like these, I immediately think: extension services. My theory is that with a half-decent module system (which we have) and a good set of functional helper structs (which I'm working on), events may become obsolete, by virtue of callbacks or specialized interfaces.

**Comment by XedinUnknown** [comment]
Interesting. I'm not sure I completely understand your perspective.

1. The Event is similar to a PSR Server Request when it comes to data. It is a mechanism to pass data from emitter to handlers, and vice-versa, and between handlers. In that way, it is similar to middleware. And when it comes to generic data, you can never promise that there _will_ be the data you require. Of course, you can always specialize your interfaces if you need to, and create specific APIs for that. Of course, data is _not_ the only thing an event have, as it also has a name, and the ability to stop its propagation.

2. Your solution can help with filtering data, but doesn't really help with one-way notifications. These need to be triggered at specific times during execution of code that is within classes, and this provide entrypoints for many things, such as logging. I don't see how extensions could _completely_ replace events. One way to approach this would be to pass the DI container instance into the class, and then doing `$c->get('my_value')` on it would be somewhat similar to triggering a `my_value` event. However, with Containers the emitter has no chance to pass any value to the `get()` method, whether it is to provide a default (the case of WP filters), or to expose data intrinsic to the emitter to 3rd parties.

**Comment by mecha** [comment]
You misunderstood me. All I'm saying is that I don't like generic events, and I can't understand why any system would rely on something so generic when invoking a callback, or better a functor, is probably (emphasis on present uncertainty) a more SOLID approach.

But this is probably not the right thread for this discussion. And my opinion shouldn't matter that much. If I _did_ like events of this kind, then this PR LGTM. But again, I'd preface that by saying that I don't care _that much_, which could have affected my judgment while reviewing. But I did do my best ;) That is all I wanted to say before.

File: composer.json [hunk 4]
```diff
@@ -1,45 +1,49 @@
...
+            "name": "Anton Ukhanev",
+            "email": "xedin.unknown@gmail.com",
+            "role": "Developer"
+        }
+    ],
+    "minimum-stability": "dev",
+    "prefer-stable": true,
+    "require": {
+        "php": "^7.0",
+        "dhii/event-dispatcher-interface": "dev-task/adapt-for-php-7.1"
```
Comment (XedinUnknown) [inline composer.json]:
```suggestion
        "dhii/event-dispatcher-interface": "^0.1"
```

**Pull request merged**